feat: post model-sync patches

ankush opened the pull request:
Summary: ability to run patches after doctype schema is migrated. `reload_doc` begone. 

Why: Because you don't need to have access to old schema always. Patch that run post-migration essentially don't ever require `reload_doc` thus eliminating the most common type of patch failure. 

Changes (some unrelated):
- [X] modify patch handler to accept type of patches to run. (pre/post)
- [X] Run all patches by default.
- [X] remove "webnotes->frappe" compatibility code. webnotes is ded. 
- [x] remove hard coded ignoring of apps called "shopping_cart"
- [x] remove "log" function that was actually just inferior print function 😄 
- [X] update migrate.py to run post-migration patches after doctype schema change. 
- [x] convert 1 remaining raw query to ORM
- [x] move few existing frappe patches to post-migration.
- [x] mark post patches as completed on install.
- [x] update docs/instructions on patching. (available as draft on docs site, will submit after merge)

Partially closes https://github.com/frappe/frappe/issues/15089 


In action (check CI run of this PR):

<img width="1020" alt="Screenshot 2021-12-17 at 8 48 53 PM" src="https://user-images.githubusercontent.com/9079960/146566590-e4a632a5-f3bc-4323-8b3f-e67af6f6abcd.png">

docs: https://frappeframework.com/compare?wiki_page=docs/v13/user/en/database-migrations&&compare=be05429005

ERPNext pr: https://github.com/frappe/erpnext/pull/28940

ankush commented:
The coverage report is sorta misleading since the patch test isn't reported 😄 but anyhoo, will add few unit test for utils.

surajshetty3416 commented on frappe/modules/patch_handler.py:
```diff
@@ -1,37 +1,58 @@
+	pass
+
+PATCH_FILES = {
+	"pre-migration": "patches.txt",
+	"post-migration": "patches_post.txt",
```
These names are technically wrong (because the execution of patch is part of migration)... isn't it `pre-model-sync` & `post-model-sync`?

ankush replied:
ah yes. The flag names right?

Any suggestion on file name too? `patches.txt` can't be changed since it's used in many apps by now, `patches_post.txt` is okay?

surajshetty3416 replied:
Would have preferred a single file for a patch... can we maintain a single JSON for a patch 😅 
something like

 **patch.json**
```JSON

{
	"pre_model_sync": [
		"frappe.patches.v13_0.rename_custom_client_script",
		"execute:frappe.reload_doc('core', 'doctype', 'role') #2017-05-23",
		"execute:frappe.reload_doc('core', 'doctype', 'user') #2017-10-27"
	],
	"post_model_sync": [
		"frappe.patches.v14_0.drop_data_import_legacy",
		"frappe.patches.v14_0.copy_mail_data #08.03.21"
	]
}
```
we can still support old `patch.txt` files?

ankush replied:
Single `patches.json` Sounds cool-er. 

1. We can keep `patches.txt` as "pre_model_sync" (current behaviour)
2. Using both `patches.txt` and `patches.json` shouldn't be allowed though. (no way for us to determine correct ordering)
3. In new apps boilerplate we will create `patches.json` by default.

hasnain2808 replied:
@surajshetty3416 @ankush 
patches.txt is supposed to be edited by hand
I think It will be much simpler to use two files instead of one JSON 
In JSON which we need to take care of indentation, quotes, commas, and brackets or it will get messy

ankush commented:
IMO we could be a bit more smarter about which patch has ran than just checking patch line in db 😅

ankush commented:
CI failing because `frappe` isn't considered a frappe app anymore by bench which confirms if directory is frappe app by checking `patches.txt` https://github.com/frappe/bench/blob/b0ad6426a848bf844bb51819b3efb9eeffbc51d9/bench/utils/__init__.py#L23

ankush commented:
<img width="1045" alt="Screenshot 2021-12-23 at 10 43 51 AM" src="https://user-images.githubusercontent.com/9079960/147191346-161fc51f-38ad-4f7a-a479-edc834fc618f.png">

all good for now. During transition, we'll have to keep this https://github.com/frappe/frappe/pull/15351/commits/3796d8fbde5b91b7723e993afd1d57b5426f3f9f to accommodate current bench users.

gavindsouza commented:
Not really excited about converting the `.txt` to `.json`. I think we should opt for a config language that's 
* Less fragile, less noise. 
* More readable for humans.
* Something that doesn't enforce us to end strings with `"` and doesn't break based on the existence of an extra trailing `,`
* Also, I'd like to not have the issue of styling come into place at all. 

How about TOML, HJSON, or something else? I'd recommend YAML too, but I don't like confusing between "3.10" and 3.10  LOL. I personally would prefer TOML for this case.

ankush commented:
patch list isn't really configuration though 👀, it's actually just list of scripts to execute in defined order... as such it's actually part of "code" 

I don't see much value in TOML here (besides adding additional deps and burden on format that many don't understand well 👀 )

JSON is shitty, but it's widely used and understood. Trailing `,` issue gets caught immediately (annoying, yet not hard to fix)


Style issues are all self-inflicted due to old decisions. Just black/prettier everything and forget about it 🥲 (github will soon support git blame ignore files) https://github.com/github/feedback/discussions/5033#discussioncomment-1549036

ankush commented:
A more robust solution could be having "posting_date" on patch files. Patches execute in the order of specified date in the files. Patches are all found in single directory (or any typical structure like `doctypefolder/patches` for doctype-specific patches. 

During migration we "discover" all patches, sort them by posting date and execute while keeping track of what's been executed already.

gavindsouza commented:
Doesn't have to be TOML. I don't see how it matters if patches are config or code. The trailing `,` might not seem so problematic here but that has bought down many Frappe production instances to date. Be it site configs or sidebar files, we've written [action](https://github.com/frappe/frappe_docs/blob/master/.github/workflows/validate_docs.yml)s to check if the JSON files are valid even ;) 

Would be so much easier to have a file that looks like 

```
[pre]
frappe.patches.v13_0.rename_custom_client_script
execute:frappe.reload_doc('core', 'doctype', 'role') #2017-05-23
execute:frappe.reload_doc('core', 'doctype', 'user') #2017-10-27

[post]
frappe.patches.v14_0.drop_data_import_legacy
frappe.patches.v14_0.copy_mail_data #08.03.21
```

instead of this bulky json format. That's all I'm saying. _(Also, that Python has standard support for files like these - https://docs.python.org/3/library/configparser.html)_

---

> A more robust solution could be having "posting_date" on patch files. Patches execute in the order of specified date in the files. Patches are all found in single directory (or any typical structure like `doctypefolder/patches` for doctype-specific patches.
> 
> During migration we "discover" all patches, sort them by posting date and execute while keeping track of what's been executed already.

And a much better solution would be to get rid of the `patches.*` file altogether. We've faced this issue quite a few times when you write a patch and it's not being run because we've forgotten to add it. But making this magical would require some more thought perhaps.

ankush commented:
let's do TOML then? (git will likely also preserve history since it's close to previous txt file)

ankush commented:
let's do ~~TOML~~ INI then? (git will likely also preserve history since it's close to previous txt file)

@surajshetty3416 @gavindsouza

ankush commented:
closing for now, will resume later

ankush commented:
go away stalebot

ankush commented:
psss: squash this while merging. Too painful to rebase and fixup now 🥲

ankush commented on frappe/patches.txt:
```diff
@@ -87,7 +88,6 @@ frappe.patches.v11_0.set_missing_creation_and_modified_value_for_user_permission
 frappe.patches.v11_0.set_default_letter_head_source
 frappe.patches.v12_0.set_primary_key_in_series
 execute:frappe.delete_doc("Page", "modules", ignore_missing=True)
-frappe.patches.v11_0.set_default_letter_head_source
```
removed cause duplicate

ankush replied:
INI parser will automatically fail when duplicates are added (which is kinda inline with how `patch log` is implemented anyway.

ankush commented on frappe/patches.txt:
```diff
@@ -123,7 +123,7 @@ frappe.patches.v12_0.remove_parent_and_parenttype_from_print_formats
 execute:from frappe.desk.page.setup_wizard.install_fixtures import update_genders;update_genders()
 frappe.patches.v12_0.set_correct_url_in_files
-execute:frappe.reload_doc('core', 'doctype', 'doctype', force=True)
+execute:frappe.reload_doc('core', 'doctype', 'doctype')
```
another duplicate; but felt required here so removed forcing (was unnecessary anyway)

ankush commented:
Failing tests *feel* unrelated; still kinda sus since these are new tests so some re-ordering of tests messed em up maybe.

ankush committed:
feat: post-model-sync patches

Ability to run few patches after doctype model schema is synced.

Read module level docstring of patch_handler.py for more info.
